File: ui/app/protected/model/ConstructionSite.ts
import { Translation } from '@protected/pages/projectgroup/queries/concept-assignment'

export interface ConstructionSite {
  constructionAreaId: string
  constructionSiteId: string
  comment: string
  fid: string
  text: string
  shape: any
}

export type ConstructionSiteKey = Pick<ConstructionSite, 'constructionAreaId' | 'constructionSiteId'>

export interface ConstructionSiteDetails {
  key: ConstructionSiteKey
  //    Bebauungsform
  form: string,
  //    Art der baulichen Nutzung
  zoningClassification: string,
  //    Maß der baulichen Nutzung
  levelOfBuiltDevelopment: string,
  //    Marktsegmente
  marketSegments: string,
  //    Energieversorgung
  energySupply: string,
  //    Mobilität
  mobility: string,
  //    Freiraum
  clearance: string,
  //    Fläche Baublock [m2]
  areaBuildingBlock: string,
  // Überbaubare Fläche [m2]
  plotAreaToBeBuiltOn: string,
  // Grundstückspreis [€/m2]
  landPricePerSqm: string
}


export type ConstructionSiteDetailsKeys = keyof Omit<ConstructionSiteDetails, 'key'>

export const ConstructionSiteDetailsKeysTranslation: Translation<ConstructionSiteDetailsKeys> = {
  form: 'Bebauungsform',
  zoningClassification: 'Art der baulichen Nutzung',
  levelOfBuiltDevelopment: 'Maß der baulichen Nutzung',
  marketSegments: 'Marktsegmente',
  energySupply: 'Energieversorgung',
  mobility: 'Mobilität',
  clearance: 'Freiraum',
  areaBuildingBlock: 'Fläche Baublock [㎡]',
  plotAreaToBeBuiltOn: 'Überbaubare Fläche [㎡]',
  landPricePerSqm: 'Grundstückspreis [€/㎡]',
}

File: src/main/kotlin/de/immovativ/vermarktungsplattform/model/ConstructionSite.kt
package de.immovativ.vermarktungsplattform.model

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonArray
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

@Serializable
data class ConstructionSiteDetails(
    val key: ConstructionSiteKey,
    //    Bebauungsform
    val form: String,
    //    Art der baulichen Nutzung
    val zoningClassification: String,
    //    Maß der baulichen Nutzung
    val levelOfBuiltDevelopment: String,
    //    Marktsegmente
    val marketSegments: String,
    //    Energieversorgung
    val energySupply: String,
    //    Mobilität
    val mobility: String,
    //    Freiraum
    val clearance: String,
    //    Fläche Baublock [m2]
    val areaBuildingBlock: String,
    // Überbaubare Fläche [m2]
    val plotAreaToBeBuiltOn: String,
    // Grundstückspreis [€/m2]
    val landPricePerSqm: String

)

@Serializable
data class ConstructionSiteKey(
    val constructionAreaId: String,
    val constructionSiteId: String
)

@Serializable
data class ConstructionSite(
    val constructionAreaId: String,
    val constructionSiteId: String,
    val comment: String,
    val fid: String,
    val text: String,
    val shape: JsonElement
)

fun List<ConstructionSite>.toFeatureCollection(): JsonObject {
    return JsonObject(
        mapOf(
            "type" to JsonPrimitive("FeatureCollection"),
            "name" to JsonPrimitive("baufelder"),
            "features" to JsonArray(map { it.shape })
        )
    )
}
